### PR TITLE
Update device profile - direct play, transcoding, and codec profiles

### LINKED ISSUE #944: Jellyfin Doesn't Transcode HEVC 4K to 1080p on Roku 4
**Software Versions**
Jellyfin Server Version: 10.8.8
Roku Client Version: 1.6 build 2
**Describe the bug**
On a Roku 4, which is 4K capable, but attached to a 1080p TV, Jellyfin transcoded HEVC 4K files to h264 4K, rather than transcoding to 1080p. Transcoding an HEVC 1080p file to h264 1080p works well. It appears that Jellyfin client isn't reporting the correct TV resolution to the server, resulting in the transcoded resolution staying unchanged. This may be related to [issue #722](https://github.com/jellyfin/jellyfin-roku/issues/722) which was marked as closed, but doesn't seem to address the issue. I don't think it's a bitrate issue, as I can play h264 1080p files that are 38Mbps (direct BD remuxes via makemkv) without any problem, so I don't trust the Roku chart mentioned in the discussion of #722.

**Expected behavior**
HEVC 4K content should be transcoded to h264 1080p content on a Roku 4 attached to a 1080p TV.

**Logs**
**Pertinent section from my transcode log:**
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> h264 (h264_nvenc))
  Stream #0:2 -> #0:1 (copy)
Press [q] to stop, [?] for help
Output #0, hls, to '/config/transcodes/27ffc89e272dd256a277151732cc2590.m3u8':
  Metadata:
    encoder         : Lavf59.27.100
  Stream #0:0: Video: h264 (High), cuda(tv, bt709, progressive), 3840x2160 [SAR 1:1 DAR 16:9], q=2-31, 98547 kb/s, 23.98 fps, 90k tbn
    Metadata:
      encoder         : Lavc59.37.100 h264_nvenc
    Side data:
      cpb: bitrate max/min/avg: 98547528/0/98547528 buffer size: 197095056 vbv_delay: N/A
  Stream #0:1: Audio: ac3, 48000 Hz, 5.1(side), fltp, 640 kb/s (default)

**Connection Information**
Server is local (LAN) and connection is http.

### LINKED ISSUE #1307: Roku app does not check device maximum resolution
## Software Versions

- Jellyfin Server Version:  10.8.10
- Roku Client Version:  1.6.6

## Describe the bug
Jellyfin client app for Roku does not check for the maximum resolution of the client, which prevents files that are above that maximum resolution from playing. Example specifically regards any Roku devices not capable of 4K, such as the 3930X.

## How To Reproduce
1. On a non-4K-capable Roku device, attempt to play any 4K video.
2. Observe as the video will not transcode down to the appropriate resolution.
3. Video will not play. "An error was encountered while playing this item"

## Expected behavior
File should automatically be transcoded to the appropriate resolution for the device, 1080p in this instance, and the media should play.

## Connection Information
Server is on local network, wireless network over an HTTP connection.

## Additional context
I discussed this issue on the Jellyfin Discord server, where Sevenrats instructed me to create this bug report. I was also informed that a log would be unnecessary in the same way, hence the lack of one.

### LINKED ISSUE #1348: Respect Maximum Transcoding Resolution user setting
## Describe the feature you'd like
When my Roku Streaming Stick+ transcodes video, it requests the transcoded video to be in the original resolution. My server is somewhat underpowered, so I'd prefer if my transcodes are never over 1080p. 

There exists a user-level setting for maximum transcode resolution, which I've set, but the Roku doesn't respect this setting.

It would also be fine to have a maximum transcode resolution setting in the Roku app itself, if that's easier to implement.

## Additional context
My TV is 1080p, but the Streaming Stick+ can play back 4K files on it just fine by downscaling on the client side. Unfortunately, HDR files need tonemapping. My underpowered server can tonemap when scaling down to 1080p at ~36 FPS, but when trying to maintain the full 4K resolution it slows down to ~16 FPS. If I had a way to tell the Roku to request the video in 1080p or lower resolution whenever it's transcoding, it would be much more convenient than trying to maintain HDR and SDR copies of my 4K content.

### PR DESCRIPTION
**Pull request by cewert**
Update the code to test each codec for each container instead of making assumptions. Updates maxAudioChannels logic - this should enable direct play support for 8 channel audio. Updates preferred audio codec if user supports surround sound. Adds a couple things from web profile i.e. `Property: "IsAnamorphic"`.

~~Still a WIP. Adding as draft~~
The main thing I still need to test is multichannel audio and I don't have hardware to test so I'm removing the draft tag. Do the codecs direct play if the device supports it? When transcoding a multichannel audio that the device can't support, which codec is used? Is the DTS user setting working? Does 8 channel audio direct play?

## Changes
- Test all containers and codecs for direct play support
  - Test for all possible video and audio codecs (except those found in user settings)
  - mpeg4 and mpeg2 user setting overrides what the device says.
- Save `maxAudioChannels` as string
  - This fixes the profile showing " 2" on occasion 
- Update `maxAudioChannels` logic to test for 8 channel support
  - This should enable direct play support for 8 channel audio
- Add support for h264 files with DolbyVision
  - docs say `Dolby Vision: dvav.09` is supported but I have no file to test with
- Add playback info to global device object
  - Height, Width, Refresh rate, Bit depth
- Create new user settings to allow the user to control the maximum resolution requested from the server.
  - Defaults: Only restrict transcoded files. Use connected TVs resolution.
  - User can choose to restrict all files instead of just trancoded files as well as being able to specify the resolution(off, auto, 360p, 480p, 720p etc.)
    - The web client setting `Maximum Allowed Video Transcoding Resolution` wasn't available using the API so I made a client side setting
- Prefer EAC3 (then AC3 -> DTS) when transcoding surround sound audio by default
  - Add user setting to prefer DTS over EAC3 and AC3
- Modify `session.user.settings.Save()` to stop trying to auto convert to int. This was causing a bug where strings were being converted to ints.
- Update session info shown in web client admin dashboard (see below for pictures)
  - Remove redundant device name and add device model
  - NOTE: The default icon shown will change to a roku icon once the 10.9.0 web client is released - https://github.com/jellyfin/jellyfin-web/pull/4104
- Add user setting to completely disable HEVC codec to improve playback for certain devices like the Ultra - #449

web client admin dashboard before
roku-session-dashboard

web client admin dashboard after
roku-session-dashboard-after

~~Built on top of PR #1287 (merge that one first)~~

Fixes #1307
Fixes #944
Fixes #1348
Addresses #904

### CODE REVIEW

**Comment by dotModded** [comment]
mpeg2 seems to be broken at line 486 of deviceCapabilities.brs, I'd fix it but I don't have the time atm:
```brightscript
mpeg2Levels = []
for each container in profileSupport
    for each level in profileSupport[container]["mpeg2"]
        if mpeg2Levels[level] = invalid
            mpeg2Levels.push(level)
        end if
    end for
end for
```
It seems as though you're treating profiles as levels for mpeg2, additionally mpeg2 profiles/levels are not numerical, see https://en.wikipedia.org/wiki/H.262/MPEG-2_Part_2#Video_profiles_and_levels I'm not sure what Roku's names for them are to test compatibility however. Not a huge issue as this format is basically dead and is more just here for completeness.

As a side note, audio transcoding with eac3 has been finnicky in my testing with a Roku Ultra. Codec support contradicts: https://developer.roku.com/docs/specs/media/streaming-specifications.md. TS/MP4 did not support E-AC3 or DTS (Even though the containers themselves support it, possibly something with the Jellyfin Server logic). 

Currently testing your changes with the if condition removed, transcoding DTS:X to a device that does not support it, but supports eac3 is still playing back in aac with your changes. My setup supports multichannel AAC however I find Jellyfin to do an extremely poor job converting it. DD/DD+ offer far better results to my ears (even though the codecs are extremely similar)

**Comment by dotModded** [comment]
Further testing writing out the profiles manually on the Roku Ultra, disabling all codecs on the playback device besides AAC, DD, and DD+. Assuming Jellyfin server prioritizes lists from first element being the most significant.

TS: 
 - Supports E-AC3, AC3, and AAC
 - AC3 is not chosen if AAC is present in any position.
 - E-AC3 is not chosen if AC3/AAC is present in any position.
 
 MP4
 - Supports AC3 and AAC
 - E-AC3 results in no audio output

MKV:
 - Supports E-AC3, AC3, and AAC
 - E-AC3 is not chosen if AC3 is present in any position.

**Comment by cewert** [comment]
@dotModded Thank you for all the feedback!

> As a side note, audio transcoding with eac3 has been finnicky in my testing with a Roku Ultra. Codec support contradicts: [developer.roku.com/docs/specs/media/streaming-specifications.md](https://developer.roku.com/docs/specs/media/streaming-specifications.md). TS/MP4 did not support E-AC3 or DTS (Even though the containers themselves support it, possibly something with the Jellyfin Server logic).

That's strange. So the device profile told the server it could play the file but audio didn't play? Did AC3 work?

> Currently testing your changes with the if condition removed, transcoding DTS:X to a device that does not support it, but supports eac3 is still playing back in aac with your changes. My setup supports multichannel AAC however I find Jellyfin to do an extremely poor job converting it. DD/DD+ offer far better results to my ears (even though the codecs are extremely similar)

Which if statement do you mean? Did the device profile correctly put eac3 first in the list of supported codecs (the device profile is printed when debugging)?

> Further testing writing out the profiles manually on the Roku Ultra, disabling all codecs on the playback device besides AAC, DD, and DD+. Assuming Jellyfin server prioritizes lists from first element being the most significant.

Yea, from my understanding, the Jellyfin server prioritizes the transcoding codecs from left to right but your results seem to indicate that's not the case. You were modifying the transcoding profile right? It seems like the solution would be to pick a preferred codec and remove the other multichannel codecs from the device profile. Would you agree?

PS: Are you on matrix? Would you mind if I PM you to chat about the codec stuff when I get around to working on it? I'm not an expert with codecs and you seem to know your stuff 😄

**Comment by cewert** [comment]
@dotModded I believe I fixed the bug with mpeg2 but I don't have a file to test with. Mpeg2 is the only codec that has supported levels you can test for but no specific profiles according to the docs [here](https://developer.roku.com/en-ca/docs/references/brightscript/interfaces/ifdeviceinfo.md#candecodevideovideo_format-as-object-as-object). 

I'm using an associative array called `profileSupport` to save what the device can support. Instead of creating a "dummy" profile array to save the mpeg2 levels under, I'm saving the levels as if they were profiles. I've added some comments to make that clear.

**Comment by dotModded** [comment]
@cewert No Matrix account, same handle on discord though. @dotmodded I'd love to chat more about this

**Comment by cewert** [comment]
I've added a max height and width to all the codec profiles. The height and width are set to the attached screen's max playback resolution (NOTE: UI resolution != playback resolution).

I tested with a H264 8bit 2160p mkv file with 6 channel EAC3 audio on a 1080p TV. Using unstable:
reason-for-transcoding-unstable

Using this branch:
reason-for-transcoding-pr

Note the extra reason for transcode and the roughly double FPS or speed of the transcode.

I think this makes sense for the default behavior. If there is a demand to override this behavior, we can always add a user setting in the future.

This change should fix #1307 and #944. Partially fixes #1348 but doesn't add support for the web client's user setting

Edit: This file is now crashing when the `playback.tryDirect.h264ProfileLevel` setting is turned on. The file fails to play with an error message then immediately crashes before I can read the message. I guess the silent fallback is failing or the new height/width settings messed it up somehow?

File: locale/en_US/translations.ts [hunk 1]
```diff
@@ -1114,10 +1114,24 @@
...
+            <extracomment>Settings Menu - Description for option</extracomment>
+        </message>
+        <message>
+            <source>Audio Codec Support</source>
+            <translation>Audio Codec Support</translation>
+            <extracomment>Settings Menu - Title of option</extracomment>
+		    <message>
+			    <source>Direct playing</source>
+			    <translation>Direct playing</translation>
+		    </message>
```
Comment (cewert) [inline locale/en_US/translations.ts]:
```suggestion
        <message>
            <source>Audio Codec Support</source>
            <translation>Audio Codec Support</translation>
            <extracomment>Settings Menu - Title of option</extracomment>
        </message>
	    <message>
		    <source>Direct playing</source>
		    <translation>Direct playing</translation>
	    </message>
```

**Comment by cewert** [comment]
It's ready for testing now. The main thing I need help testing is multichannel audio:

- Does 8 channel audio direct play?
- Do the codecs direct play if the device supports it? EAC3 AC3 DTS
- When transcoding a multichannel audio that the device can't support, which codec is used? EAC3 by default
- Is the DTS user setting working for transcoded files?

**Comment by cewert** [comment]
I figured out how to restrict playback resolution for only transcoded files 🎉 

I've added a user setting that allows the user to restrict resolution for all files or just transcoded files. Defaults to just transcoded files.
The other Maximum Resolution setting remains and is applied to either the Codec Profile or TranscodingProfile depending on what mode is set in settings (but only when the max res setting <> "off")

NOTE: The transcoding reason "The videos' resolution is not supported" will only show up when the mode is set to `everything`.  `transcoding` mode will not show a transcoding reason but can be observed by checking the ffmpeg log.

**Review by jimdogx — APPROVED** [review]
LGTM.  

* I tested DTS 5.1 on a non DTS system and things transcoded as expected.  Then I played the same video on a DTS capable system and it directed played.
* I also tested the Max Res setting.  Transcoded to the max resolution when the setting was on.  Then direct played when it was off.

**Review by neilsb — APPROVED** [review]
As mentioned, I only have stereo audio, so not specifically able to to test the DTS stuff etc, but used the version over the weekend and certainly didn't hit any problems, and everything appeared to work as expected.  And code all looks good.

**Pull request merged**